### PR TITLE
fix: change check to disallow optimistic sends

### PR DESCRIPTION
**Pull request by charleenfei**
## Description

closes: #2670 

### Commit Message / Changelog Entry

```bash
fix(statemachine)!: change check to disallow optimistic sends
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/keeper/packet.go [hunk 1]
```diff
@@ -33,7 +33,7 @@ func (k Keeper) SendPacket(
 		return 0, sdkerrors.Wrap(types.ErrChannelNotFound, sourceChannel)
 	}
 
-	if channel.State == types.CLOSED {
+	if channel.State != types.OPEN {
 		return 0, sdkerrors.Wrapf(
 			types.ErrInvalidChannelState,
 			"channel is CLOSED (got %s)", channel.State.String(),
```
Comment (crodriguezvega) [inline modules/core/04-channel/keeper/packet.go:39]: Should we maybe change the error text to something like `channel is not OPEN...`?

File: modules/core/04-channel/keeper/packet.go [hunk 2]
```diff
@@ -33,7 +33,7 @@ func (k Keeper) SendPacket(
 		return 0, sdkerrors.Wrap(types.ErrChannelNotFound, sourceChannel)
 	}
 
-	if channel.State == types.CLOSED {
+	if channel.State != types.OPEN {
```
Comment (crodriguezvega) [inline modules/core/04-channel/keeper/packet.go:36]: Should we add a test like [this one](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/keeper/packet_test.go#L98-L104), but for `TRY_OPEN`?

**Review by crodriguezvega — COMMENTED** [review]
Thank you, @charleenfei! I left two suggestions.

**Review by AdityaSripal — APPROVED** [review]

File: testing/endpoint.go [hunk 3]
```diff
@@ -566,11 +566,11 @@ func (endpoint *Endpoint) TimeoutOnClose(packet channeltypes.Packet) error {
 	return endpoint.Chain.sendMsgs(timeoutOnCloseMsg)
 }
 
-// SetChannelClosed sets a channel state to CLOSED.
-func (endpoint *Endpoint) SetChannelClosed() error {
+// SetChannelState sets a channel state
+func (endpoint *Endpoint) SetChannelState(state channeltypes.State) error {
```
Comment (damiannolan) [inline testing/endpoint.go:570]: Nice! We could avoid the api breaking change here in the testing package by just adding a new function `SetChannelState(state)` and keep `SetChannelClosed` as well.

**Review by damiannolan — APPROVED** [review]

File: testing/endpoint.go (hunk 3 above)
Comment (crodriguezvega) [reply testing/endpoint.go:570]: I like this refactoring. Should we anyway break the API and document this in the migration docs for v7?

File: modules/core/04-channel/keeper/packet_test.go [hunk 4]
```diff
@@ -95,11 +95,25 @@ func (suite *KeeperTestSuite) TestSendPacket() {
 			sourceChannel = ibctesting.InvalidID
 			channelCap = suite.chainA.GetChannelCapability(path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
 		}, false},
-		{"channel closed", func() {
+		{"channel is in CLOSED stage", func() {
```
Comment (crodriguezvega) [inline modules/core/04-channel/keeper/packet_test.go:98]:
```suggestion
		{"channel is in CLOSED state", func() {
```

File: modules/core/04-channel/keeper/packet_test.go [hunk 5]
```diff
@@ -95,11 +95,25 @@ func (suite *KeeperTestSuite) TestSendPacket() {
...
 			suite.coordinator.Setup(path)
 			sourceChannel = path.EndpointA.ChannelID
 
-			err := path.EndpointA.SetChannelClosed()
+			err := path.EndpointA.SetChannelState(types.CLOSED)
+			suite.Require().NoError(err)
+		}, false},
+		{"channel is in INIT stage", func() {
```
Comment (crodriguezvega) [inline modules/core/04-channel/keeper/packet_test.go:105]:
```suggestion
		{"channel is in INIT state", func() {
```

File: modules/core/04-channel/keeper/packet_test.go [hunk 6]
```diff
@@ -95,11 +95,25 @@ func (suite *KeeperTestSuite) TestSendPacket() {
...
+			suite.coordinator.Setup(path)
+			sourceChannel = path.EndpointA.ChannelID
+
+			err := path.EndpointA.SetChannelState(types.INIT)
+			suite.Require().NoError(err)
+		}, false},
+		{"channel is in TRYOPEN stage", func() {
```
Comment (crodriguezvega) [inline modules/core/04-channel/keeper/packet_test.go:112]:
```suggestion
		{"channel is in TRYOPEN state", func() {
```

**Review by crodriguezvega — COMMENTED** [review]
Nice work, @charleenfei! If we agree to break the API for the testing package, then we only need to mention that in the migration docs.

File: testing/endpoint.go (hunk 3 above)
Comment (charleenfei) [reply testing/endpoint.go:570]: I agree with @crodriguezvega that breaking the API and updating migration docs in this case is cleaner...do you have a strong opinion about not breaking it @damiannolan ?

**Pull request merged**